GSA/searchgov-spider
# SRCH-5430 Conditional JS Spider

## Summary
#### Brief summary of the changes included in this PR
* Functional Code Changes
  * Adds the ability to either use or not use playwright.  Initially this was combined into the `domain_spider` but given unknowns about how we want to use these, splitting them out into `domain_spider` and `domain_spider_js` seems like it gives us the most flexibility to keep it separate or merging back into a single spider later.  Having a single spider I could not get it to accept the "full" list of domains/urls and within that list differentiate between those needing javascript and not.  I could pass in a list of those needing javascript and then a list of those not needing javascript to the same spider but separate spiders maybe makes it more clear what you mean to do.
  * Moved the helper variables and functions into a helpers module since they were just referencing each other back and forth it seems simplest to place them in a single file.  I also moved some other common things
  * I removed the. manual duplication check going on with `url_map`.  We have a `unique=True` in the link extractor and a de-duplication item pipeline.  The `url_map` functionality did not seem to be doing anything, i saw no differences in url output when i removed it.
  * Replaced `domains.txt` and `startingUrls.txt` with `crawl-sites.json` so that we can apply more attributes to search crawl.  The spiders are setup to filter only those that match their javascript needs from this file.  I also wrote a script that generates this file from a scrutiny plist file in case we need to use it again.
  * Updated the readme with references to `domain_spider_js`

* Codebase improvements
  * Replaced `.placeholder` files with `.gitignore` so that we can also ignore all the files that get created in those directory since I'm assuming we don't ever want to check those in.  I also modified the top level `.gitignore` to ignore 
  * Simplified the list of dependencies in the `requirements.txt` file (and `pyproject.toml`) 
  * Ran `pre-commit run -a` which added missing new lines to the end of many files.

* Testing and Linting
  * Moved the tests folder out of the "src" folder so that we can have one folder for tests if we ever want to test things that aren't in the `search_gov_spiders` we can do so in one place, share fixtures, etc...  Also, if we were to build the project as a package (using `build`, etc..) we don't need to include the tests.
  * Added `betamax` to the suite of testing libraries.  I wanted to be able to run the crawler on the real site as a test.  Betamax records the requests and responses and next time it runs does not actually call out to the website but uses the contents of the `tests/search_gov_spiders/responses` as a source.  I ultimately could not get it working with the "full" crawl test which is why you see some of it commented out in the test and why I am excluding it from running in circle.  Ultimately maybe this is not the right solution but I really wanted a way to make sure my changes and future changes were not breaking something.
  * Even though pylint is disabled in pre-commit I found it raised a lot of "cannot import module" problems when i ran it seemly due to the fact that it could not access my venv.  Adding pylint to the dependencies and referencing the local version in pre-commit solved this.  I also added some configs to the `pyproject.toml` file for pylint for when we do get it added to pre-commit.
 

#### Any additional information or context which may help the reviewer
* Seems like more than enough above 🤣 

#### Instructions on How to Manually Test changes
* Source venv and install dependencies
  * `python -m pip install -r search_gov_crawler/requirements.txt`
  * Perhaps also `playwright install`
* Change to crawler root
  * `cd search_gov_crawler`
* Run domain_spider on sample site
  * `scrapy crawl domain_spider -a allowed_domains=quotes.toscrape.com -a start_urls=https://quotes.toscrape.com/`
* Run domain_spider_js on sample site
  * `scrapy crawl domain_spider_js -a allowed_domains=quotes.toscrape.com -a start_urls=https://quotes.toscrape.com/js`
* Manually verify contents of `crawl-sites.json` matches the contents of the deleted text files.  I made a few small changes even after the script ran to generate it but hopefully those make sense. 
* You can also test the script by downloading the scruity file from the JIRA story and assuming its placed in the project root you can run `python search_gov_crawler/search_gov_spiders/utility_files/import_plist.py --input_file
 ./scrutiny-2023-06-20.plist` which should overwrite the exsting `crawl-sites.json` file with the output of this run.
* it also still works with scrapyd (assuming scrapyd is running):
  * `scrapyd-deploy local`
  * `curl http://localhost:6800/schedule.json -d project=search_gov_spiders -d spider=domain_spider -d allowed_domains=quotes.toscrape.com -d start_urls=https://quotes.toscrape.com/`
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks
 
- [X] Automated checks pass, if applicable. If Code Climate checks do not pass, explain reason for failures:  #### Process Checks
